File: backend/app/services/knowledge_base.py
import json
from pathlib import Path
from typing import Any

from backend.app.config import settings


def _read_json(path: Path) -> dict[str, Any]:
    with path.open("r", encoding="utf-8") as file:
        return json.load(file)


class KnowledgeBase:
    """Loads a topic pack used by the agents."""

    def __init__(self, topic_id: str, base_dir: Path | None = None) -> None:
        self.topic_id = topic_id
        root = base_dir or settings.knowledge_base_dir
        self.base_dir = root / "topics" / topic_id
        self.manifest = _read_json(self.base_dir / "manifest.json")
        self.concepts = _read_json(self.base_dir / "concept_graph.json")
        self.misconceptions = _read_json(self.base_dir / "misconceptions.json")
        self.interventions = _read_json(self.base_dir / "interventions.json")
        self.problems = _read_json(self.base_dir / "problems.json")
        self.evaluation_rules = _read_json(self.base_dir / "evaluation_rules.json")

    def find_misconception(self, concept: str, error_tag: str | None) -> dict[str, Any] | None:
        for item in self.misconceptions["items"]:
            if item["concept_id"] == concept and error_tag in item.get("error_tags", []):
                return item
        return None

    def find_interventions(self, concept: str, error_tag: str | None) -> list[dict[str, Any]]:
        matches = []
        for item in self.interventions["rules"]:
            if item["concept_id"] != concept:
                continue
            misconception = self.find_misconception(concept, error_tag)
            if misconception and item["misconception_id"] == misconception["id"]:
                matches.append(item)
        return matches

File: backend-node/src/services/knowledgeBase.js
/**
 * Knowledge base service — loads a topic pack used by the agents.
 *
 * Equivalent of backend/app/services/knowledge_base.py
 */

import fs from "fs";
import path from "path";
import settings from "../config.js";

function readJson(filePath) {
  const raw = fs.readFileSync(filePath, "utf-8");
  return JSON.parse(raw);
}

export class KnowledgeBase {
  /**
   * @param {string} topicId  - The snake_case topic identifier
   * @param {string} [baseDir] - Override for the knowledge_base directory
   */
  constructor(topicId, baseDir = null) {
    this.topicId = topicId;
    const root = baseDir || settings.knowledgeBaseDir;
    this.baseDir = path.join(root, "topics", topicId);

    if (!fs.existsSync(this.baseDir)) {
      throw Object.assign(
        new Error(`Topic pack not found: ${this.baseDir}`),
        { status: 404 }
      );
    }

    this.manifest = readJson(path.join(this.baseDir, "manifest.json"));
    this.concepts = readJson(path.join(this.baseDir, "concept_graph.json"));
    this.misconceptions = readJson(path.join(this.baseDir, "misconceptions.json"));
    this.interventions = readJson(path.join(this.baseDir, "interventions.json"));
    this.problems = readJson(path.join(this.baseDir, "problems.json"));
    this.evaluationRules = readJson(path.join(this.baseDir, "evaluation_rules.json"));
  }

  /**
   * Find a misconception entry matching a concept and error tag.
   */
  findMisconception(concept, errorTag) {
    for (const item of this.misconceptions.items || []) {
      if (
        item.concept_id === concept &&
        (item.error_tags || []).includes(errorTag)
      ) {
        return item;
      }
    }
    return null;
  }

  /**
   * Find all intervention rules matching a concept and error tag.
   */
  findInterventions(concept, errorTag) {
    const matches = [];
    for (const rule of this.interventions.rules || []) {
      if (rule.concept_id !== concept) continue;
      const misconception = this.findMisconception(concept, errorTag);
      if (misconception && rule.misconception_id === misconception.id) {
        matches.push(rule);
      }
    }
    return matches;
  }
}
